### DISCUSSION TITLE
A question about how to best build a relalation between the ECS system and godots physics system

### DISCUSSION THREAD

**Discussion by mk56-spn** [friflo/Friflo.Engine.ECS#39] [Ideas]
Currently I am developing a game in Godot and using this ecs as a backbone since I'm not uses nodes basically at all 

However I have one question. are unique entities expensive ? 

I ask this because I am developing an editor and find myself needing to query the physics system in Godot for the placement of objects. Which is fine. 

But then retrieving the corresponding entity from friflo has made me have to pass in a string ( which I cast from godots  physics RID ) and attach a unique id to all my entities. 

The number of entities I'm looking at having to manage is around 30-100 k at most . Not run into any issues so far but I just wanted to make sure I'm not abusing a system in a problematic way. 

And thank you for developing this excellent ecs !

**Comment by friflo** [comment]
I have no deep experience with Godot.
From the infos you gave I searched at their GitHub Repo.
If found the struct [Rid](https://github.com/godotengine/godot/blob/master/modules/mono/glue/GodotSharp/GodotSharp/Core/Rid.cs).

So I assume you are using this type for physics.
I did this because you are mentioned this:

> But then retrieving the corresponding entity from friflo has made me have to pass in a string ( which I cast from godots physics RID ) and attach a unique id to all my entities.

Mapping & converting to/from a `string` is something I would avoid if I have a focus on performance.

The `Rid` struct above uses:
```cs
   public readonly struct Rid : IEquatable<Rid>
   {
        private readonly ulong _id; // Default is 0
```
Using `ulong` instead of a string is nice. Lookup/ mapping of this type is pretty fast.
Using a `string` for lookup/mapping/serialization/... is significant slower.

Not sure if this helps.

**Reply by mk56-spn** [reply]
Thanks for the fast response. I'll test this when I get home. I assume by your recommendation that uniqueentity allows a ulong to be used then as it's unique identifier? . If not then it's fine I'll use a key pair dictionary to track the corresponding rid
 

And yes that's the rid I was talking about. It's what Godot provides you with when you query it for objects that are in aj area or pijt for example. I imagine it's a similar thing  in most engines

**Comment by friflo** [comment]
I assume use mean the [UniqueEntity](https://friflo.gitbook.io/friflo.engine.ecs/examples/general#unique-entity) described in the docs.

Since 3.0.0-preview version there is a more performant and flexible way to lookup entities by a custom type.
In your case it would look something like.

```
struct RidComponent: IIndexedComponent<ulong>
{
    public  ulong rid;
    public  ulong GetIndexedValue() => rid;  // indexed field
}
```

An ` IIndexedComponent<>` behaves like a component.
Additionally it supports the lookup of entities having a component with a specific value.
In your case the `ulong` of an `Rid`.

More info at [Search](https://friflo.gitbook.io/friflo.engine.ecs/examples/component-types#search)

**Reply by mk56-spn** [reply]
I somehow missed this. Excellent and fills the role I need perfectly. Sorry for wasting your time. Perhaps it could be moved to the docs parts where unique entity is after 3.0 goes live ?

**Comment by friflo** [comment]
Reopen discussion. Reasons:
- To have a use-case for indexed components in **Discussions**
- To add note to use indexed components instead of `UniqueEntity`'s https://friflo.gitbook.io/friflo.engine.ecs/examples/general#unique-entity docs as you suggested
